llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java

```diff
@@ -352,18 +440,21 @@ public void allocateMultiple(MemoryBuffer[] dest, int size, BufferObjectFactory
               LlapIoImpl.LOG.info("Failed to deallocate after a partially successful allocate: " + dest[i]);
             }
           }
+          //Need to release the un-allocated but reserved memory.
+          memoryManager.releaseMemory(dest.length - destAllocIx << allocLog2);
           String msg = "Failed to allocate " + size + "; at " + destAllocIx + " out of "
               + dest.length + " (entire cache is fragmented and locked, or an internal issue)";
           logOomErrorMessage(msg);
           throw new AllocatorOutOfMemoryException(msg);
         }
-        ++attempt;
+        ++discardsAttempt;
       }
     } finally {
+      //@TODO CHECK i think it is a bug, we do not have to release this memory if we already did deallocate(dest[i])
```

Comment:
The deallocate path doesn't increment memoryForceReleased, so this here should be fine.